### PR TITLE
[fix][fsdp] Resume bitsandbytes checkpoints strictly

### PR DESCRIPTION
**Pull request by bvolpato**
## Summary

FSDP checkpoints for bitsandbytes 4-bit modules include quantization metadata beside each packed weight. Freshly constructed modules can accept the weight while reporting that metadata as unexpected, causing strict resume to fail before optimizer, scheduler, RNG, or dataloader state is restored.

Accept known FP4/NF4 metadata only when its associated weight loaded and the metadata matches the initialized model. A mismatch must fail: `Linear4bit.load_state_dict` does not restore that quantization state, so loading NF4 packed bytes into an FP4 model otherwise silently produces wrong outputs. The same check covers changed base scales and double-quantization settings.

Missing model keys and unrelated unexpected keys still fail strict loading. Non-strict loading and the saved checkpoint format are unchanged.

## Validation

- 19 focused tests pass with actual bitsandbytes 0.50.2 and PyTorch 2.11 on CPU. The five new mismatch rejections fail against the previous helper.
- NF4/FP4 and both double-quantization settings restore exact outputs after destination packed weights are deliberately cleared.
- Independent review reran all 19 tests and verified a real nested PEFT + bitsandbytes checkpoint round trip, including trained adapters.
- Without the optional bitsandbytes dependency, 12 tests pass and 7 integration cases skip.
- Required Ruff, Black, gitleaks, and diff checks pass.

The earlier PR version had single-GPU NF4 QLoRA resume validation from step 20 to 40 and then 40 to 60, with training state restored, nonzero gradients, and subsequent checkpoints. The new metadata comparison has CPU regression coverage; GPU/FSDP integration has not been rerun for this follow-up.

## Repro command

```bash
uv run --isolated --extra fsdp --extra dev pytest --noconftest \
  tests/backends/skyrl_train/distributed/test_fsdp_checkpoint.py -q
```

## Downsides

Strict loading compares ignored scale and codebook tensors on the CPU, adding host copies when initialized quantization state is on the GPU. It does not gather packed model weights or add distributed collectives. Resume requires the same base model and quantization settings. As with PyTorch strict loading, an exception can occur after some parameter values were copied; callers must treat that restore as failed.

### CODE REVIEW

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 1]
```diff
@@ -56,6 +56,39 @@
...
+    ".nested_quant_map",
+    ".quant_state.bitsandbytes__fp4",
+    ".quant_state.bitsandbytes__nf4",
+)
+
+
+def _is_bnb_quantization_state_key(key: str, loaded_keys: set[str]) -> bool:
+    return any(
+        key.endswith(suffix) and key[: -len(suffix)] in loaded_keys for suffix in _BNB_QUANTIZATION_STATE_SUFFIXES
+    )
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/distributed/fsdp_strategy.py]: Applied in e283a483. Added tuple precheck before suffix matching. Focused checkpoint suite: 9 passed; changed-file Ruff, Black, and gitleaks passed.

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 2]
```diff
@@ -56,6 +56,39 @@
...
+
+
+def _load_model_state_dict(model: nn.Module, state_dict, *, strict: bool) -> None:
+    if not strict:
+        model.load_state_dict(state_dict, strict=False)
+        return
+
+    incompatible = model.load_state_dict(state_dict, strict=False)
+    loaded_keys = set(state_dict).difference(incompatible.unexpected_keys)
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/distributed/fsdp_strategy.py]: Applied in e283a483. `state_dict.keys()` now performs subtraction directly. Focused checkpoint suite: 9 passed; changed-file Ruff, Black, and gitleaks passed.

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 3]
```diff
@@ -56,6 +56,42 @@
...
+    return False
+
+
+def _load_model_state_dict(model: nn.Module, state_dict, *, strict: bool) -> None:
+    if not strict:
+        model.load_state_dict(state_dict, strict=False)
+        return
+
+    incompatible = model.load_state_dict(state_dict, strict=False)
+    loaded_keys = state_dict.keys() - incompatible.unexpected_keys
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/distributed/fsdp_strategy.py:84]: This is not reproducible on supported Python. Both `dict_keys` and `odict_keys` accept list operands for subtraction and return a set: `{'a': 1, 'b': 2}.keys() - ['b'] == {'a'}`. Focused strict-load tests also pass with PyTorch's list-valued `unexpected_keys`.

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 4]
```diff
@@ -56,6 +56,42 @@
...
+)
+
+
+def _is_bnb_quantization_state_key(key: str, loaded_keys: set[str]) -> bool:
+    if not key.endswith(_BNB_QUANTIZATION_STATE_SUFFIXES):
+        return False
+    for suffix in _BNB_QUANTIZATION_STATE_SUFFIXES:
+        if key.endswith(suffix):
+            return key[: -len(suffix)] in loaded_keys
+    return False
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/distributed/fsdp_strategy.py:75]: These suffixes do not overlap because their delimiters differ: `weight.nested_absmax`.endswith(`.absmax`) and `weight.nested_quant_map`.endswith(`.quant_map`) are both false. Focused tests cover all six supported suffixes, including both nested forms, and strict loading passes.
